src/trace/implementations/merge_batcher_col.rs

```diff
@@ -0,0 +1,353 @@
+        }
+        if !kept.is_empty() {
+            self.sorter.push_list(kept);
+        }
+
+        // Drain buffers (fast reclamation).
+        self.sorter.clear_stash();
+
+        let seal = builder.done(self.lower.clone(), upper.clone(), Antichain::from_elem(<B::Time as timely::progress::Timestamp>::minimum()));
+        self.lower = upper;
+        seal
+    }
+
+    // the frontier of elements remaining after the most recent call to `self.seal`.
+    fn frontier(&mut self) -> timely::progress::frontier::AntichainRef<B::Time> {
+        self.frontier.borrow()
+    }
+}
+
+struct TimelyStackQueue<T: Columnation> {
+    list: TimelyStack<T>,
+    head: usize,
+}
+
+impl<T: Columnation + 'static> TimelyStackQueue<T> {
+
+    fn new() -> Self { TimelyStackQueue::from(Default::default()) }
+
+    fn pop(&mut self) -> &T {
+        self.head += 1;
+        &self.list[self.head - 1]
+    }
+
+    fn peek(&self) -> &T {
+        &self.list[self.head]
+    }
+
+    fn from(list: TimelyStack<T>) -> Self {
+        TimelyStackQueue {
+            list,
+            head: 0,
+        }
+    }
+
+    fn done(mut self) -> TimelyStack<T> {
+        self.list.clear();
+        self.list
+    }
+
+    fn len(&self) -> usize { self.list.len() - self.head }
+
+    fn is_empty(&self) -> bool { self.head == self.list.len() }
+
+    /// Return an iterator over the remaining elements.
+    fn iter(&self) -> impl Iterator<Item=&T> + Clone + ExactSizeIterator {
+        self.list.iter().skip(self.head)
+    }
+}
+
+struct MergeSorterColumnation<D: Ord+Columnation, T: Ord+Columnation, R: Semigroup+Columnation> {
+    queue: Vec<Vec<TimelyStack<(D, T, R)>>>,    // each power-of-two length list of allocations.
+    stash: Vec<TimelyStack<(D, T, R)>>,
+    pending: Vec<(D, T, R)>,
+}
+
+impl<D: Ord+Clone+Columnation+'static, T: Ord+Clone+Columnation+'static, R: Semigroup+Columnation+'static> MergeSorterColumnation<D, T, R> {
+
+    const BUFFER_SIZE_BYTES: usize = 64 << 10;
+
+    /// Buffer size (number of elements) to use for new/empty buffers.
+    const fn buffer_size() -> usize {
+        let size = std::mem::size_of::<(D, T, R)>();
+        if size == 0 {
+            Self::BUFFER_SIZE_BYTES
+        } else if size <= Self::BUFFER_SIZE_BYTES {
+            Self::BUFFER_SIZE_BYTES / size
+        } else {
+            1
+        }
+    }
+
+    fn new() -> Self {
+        Self {
+            queue: Vec::new(),
+            stash: Vec::new(),
+            pending: Vec::new()
+        }
+    }
+
+    fn empty(&mut self) -> TimelyStack<(D, T, R)> {
+        self.stash.pop().unwrap_or_else(|| TimelyStack::with_capacity(Self::buffer_size()))
+    }
+
+    /// Remove all elements from the stash.
+    fn clear_stash(&mut self) {
+        self.stash.clear();
+    }
+
+    /// Insert an empty buffer into the stash. Panics if the buffer is not empty.
+    fn recycle(&mut self, buffer: TimelyStack<(D, T, R)>) {
+        assert!(buffer.is_empty());
```

Comment:
Mentioned elsewhere, but this might be a time to call `buffer.clear()` instead, as 1. it should be cheap if empty, 2. we don't want to retain the regions if it is empty but not clear, and 3. avoids a panic booby trap.

Reply:
This now check that the capacity matches the expected capacity, clears and stashes, or drops it otherwise. No more panics.